### ISSUE TITLE
Make it possible for TextInput to emit a message when gaining/losing foccus.

### ISSUE THREAD

**Issue by thenlevy** [hecrj/iced#819]
This could be useful when using iced with an existing wgpu application like in the `integration` example.

Imagine that the wgpu application is supposed to react to some keyboard inputs (for example  W A S D to move the camera) and that the iced GUI on top of it has some `TextInput`.

It would be useful to know in the main loop if of the TextInput is focused so that, in that case the keyboard events would not be forwarded to the wgpu app but only to the iced GUI.
Currently this can be achieved by asking the iced GUI, to check the state of all of its `TextInput` but I think emitting a message would make it easier.

**Comment by hecrj** [comment]
If you want to know if a specific `Event` was captured by `iced`, you can always use [`UserInterface::update`](https://docs.rs/iced_native/0.4.0/iced_native/struct.UserInterface.html#method.update) directly which returns a list of `event::Status`.

**Comment by kaimast** [comment]
How about a widget that that behaves like `Container` but allows to handle events that are not processed by its children? 

Using `UserInterface::update` works but seems to require re-implementing some of the functionality of `program::State`.

**Comment by kaimast** [comment]
So I thought about this some more: An application, for example, a game, might want to keep track of which keys are currently pressed. If a `TextInput` gains focus it should be equivalent to all keys being un-pressed/released. 

So I think its not sufficient to just forward the un-captured events to the application's input management. I think something like a `TextInputFocused` (or more generally a `UiFocused` event) might be super useful.
I don't think you need to propagate an unfocused event as well, but it might be necessary for other use cases.

(Please correct me, if I am wrong)

**Comment by ModProg** [comment]
I had a problem probably related to this, I wanted to close a window on pressing ESC but when the text input was focused the update event loop did not get the KeyPressed event. 

I did find a workaround by listening to the KeyReleased event though.

**Comment by kaimast** [comment]
For me the problem has been resolved because https://github.com/iced-rs/iced/pull/1230 was merged.

**Comment by kaimast** [comment]
> I had a problem probably related to this, I wanted to close a window on pressing ESC but when the text input was focused the update event loop did not get the KeyPressed event.
> 
> I did find a workaround by listening to the KeyReleased event though.

Seems quite different from this issue. You seem to always want to receive the ESC event independent of whether a text input is focused, or do I misunderstand?

**Comment by ModProg** [comment]
> > I had a problem probably related to this, I wanted to close a window on pressing ESC but when the text input was focused the update event loop did not get the KeyPressed event.
> > I did find a workaround by listening to the KeyReleased event though.
> 
> Seems quite different from this issue. You seem to always want to receive the ESC event independent of whether a text input is focused, or do I misunderstand?

yes, but as the textinput is the only thing present, listening for it loosing focus is also fine.
